inc/namespace.php

```diff
@@ -56,17 +56,34 @@ function get_merged_config() : array {
 
 	// Get custom config overrides.
 	$composer_file = ROOT_DIR . '/composer.json';
-	$composer_json = get_json_file_contents_as_array( $composer_file );
 
+	$composer_json = get_json_file_contents_as_array( $composer_file );
+	$environment = get_environment_type();
+	validate_config_settings( $composer_json, $environment );
```

Comment:
What was the thinking behind passing in the full `$composer_json` data? Seems we only need the data in `$config` to me

I'm thinking we should do the validation in 2 places:

1. Validate the default config _before_ the environment overrides are merged in (as you are doing)
2. Validate the config again after merging environment type overrides

This way we can direct people to where the problem is a little better for example "There is a problem in the default config" or "There is a problem in the staging config".

I think we can do the per environment validation within the `validate_config_settings()` function itself eg:

```php
function validate_config_settings( array $config ) {
     $environment = get_environment_type();

     // Get configs to validate.
     $configs = [ 
         'default' => $config
     ];

     // Validate all environment types locally to help avoid pushing unforeseen problems to production.
     if ( $environment === 'local' ) {
          $configs['local'] = merge_config_settings( $config, $config['environments']['local'] ?? [] );
          $configs['development'] = merge_config_settings( $config, $config['environments']['development'] ?? [] );
          $configs['staging'] = merge_config_settings( $config, $config['environments']['staging'] ?? [] );
          $configs['production'] = merge_config_settings( $config, $config['environments']['production'] ?? [] );
     } else {
          $configs[ $environment ] = merge_config_settings( $config, $config['environments'][ $environment ] ?? [] );
     }

     foreach ( $configs as $type => $conf ) {
         // Do config validation here, adding the config type name to the error messages.
     }
}
```

Reply:
So I finally was able to sit down and look at this and I have a few questions. First of all - I've done what @roborourke suggested, however I am not sure it couldn’t be easier. Merging all the environment configs and validating after this seems a bit overkill - the code will actually only merge the modules from each config so we can check these without the need of merging and then looping through them. That said, I don't see a better way to determine if there is module under the environment config that is not under `modules` property. I am open to suggestions for these, I might be missing something.

Reply:
I'm only suggesting doing that on local so you can see any issues. The reason is that the config overrides for a given environment might not have issues once merged with the base config so we should only validate after merging. I don't think it's overkill in that case if it prevents errors getting to production stacks.

Reply:
@roborourke  I am not saying that it's overkill to do it, just wondering if it's worth doing it this way considering that it seems only modules get merged. In any case, if you think this is best, I am ok with it.

Reply:
Right sorry, so that's not entirely the case - if you have a top-level config item it will be overridden by any matching top level items in an environment specific config. Modules are a special case as they're merged one level deeper. We'll improve this over time but currently pretty much all config that needs to be merged and checked is in the modules section